pkg/controllers/snapshot_controller.go

```diff
@@ -0,0 +1,394 @@
+			return fmt.Errorf("failed to fetch snapshot from store: %w", err)
+		}
+
+		return nil
+	}
+
+	if snapshot.DeletedAt != nil {
+		if err := r.deleteSnapshot(ctx, log, ioCtx, snapshot); err != nil {
+			return fmt.Errorf("failed to delete snapshot: %w", err)
+		}
+	}
+
+	if snapshot.Status.State == api.SnapshotStatePopulated {
+		log.V(1).Info("Snapshot already populated")
+		return nil
+	}
+
+	snapshot.Finalizers = append(snapshot.Finalizers, SnapshotFinalizer)
+	if _, err := r.store.Update(ctx, snapshot); err != nil {
+		return fmt.Errorf("failed to set finalizers: %w", err)
+	}
+
+	rc, size, digest, err := r.openSnapshotSource(ctx, snapshot.Source)
+	if err != nil {
+		return fmt.Errorf("failed to open snapshot source: %w", err)
+	}
+	defer func() {
+		if err := rc.Close(); err != nil {
+			log.Error(err, "failed to close snapshot source")
+		}
+	}()
+
+	options := librbd.NewRbdImageOptions()
+	defer options.Destroy()
+
+	//TODO: different pool for OS images?
+	if err := options.SetString(librbd.RbdImageOptionDataPool, r.pool); err != nil {
+		return fmt.Errorf("failed to set data pool: %w", err)
+	}
+	log.V(2).Info("Configured pool", "pool", r.pool)
+
+	roundedSize := round.OffBytes(size)
+	if err = librbd.CreateImage(ioCtx, SnapshotIDToRBDID(snapshot.ID), roundedSize, options); err != nil {
+		return fmt.Errorf("failed to create os rbd image: %w", err)
+	}
+	log.V(2).Info("Created rbd image", "bytes", roundedSize)
+
+	rbdImg, err := librbd.OpenImage(ioCtx, SnapshotIDToRBDID(snapshot.ID), librbd.NoSnapshot)
+	if err != nil {
+		return fmt.Errorf("failed to open rbd image: %w", err)
+	}
+
+	if err := r.populateImage(log, rbdImg, rc); err != nil {
+		return fmt.Errorf("failed to populate os image: %w", err)
+	}
+	log.V(2).Info("Populated os image on rbd image")
+
+	imgSnap, err := rbdImg.CreateSnapshot(ImageSnapshotVersion)
+	if err != nil {
+		return fmt.Errorf("unable to create snapshot: %w", err)
+	}
+
+	if err := imgSnap.Protect(); err != nil {
+		return fmt.Errorf("unable to protect snapshot: %w", err)
+	}
+
+	snapshot.Status.Digest = digest
+	snapshot.Status.State = api.SnapshotStatePopulated
+
+	if _, err = r.store.Update(ctx, snapshot); err != nil {
+		return fmt.Errorf("failed to update snapshot metadate: %w", err)
+	}
+
+	return nil
+}
+
+func (r *SnapshotReconciler) populateImage(log logr.Logger, dst io.WriteCloser, src io.Reader) error {
+	rater := utils.NewRater(src)
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	done := make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				log.Info("Populating", "rate", rater.String())
+			case <-done:
+				return
+			}
+		}
+	}()
+	defer func() { close(done) }()
+
+	buffer := make([]byte, r.populatorBufferSize)
+	_, err := io.CopyBuffer(dst, rater, buffer)
+	if err != nil {
+		_ = dst.Close()
+		return fmt.Errorf("failed to populate image: %w", err)
+	}
+	log.Info("Successfully populated image")
+
+	return nil
+}
```

Comment:
`dst` is not closed on success. Also, `populateImage` should not call `dst.Close` in the first place, as it doesn't actually manage the `io.WriteCloser` since the calling method is the entity opening / having to close the image.
In general, for operations involving `io.Closer`, either you return the `io.Closer` or the function opening it should be the one closing it, which means `Close()` should be called in `reconcileSnapshot`.